api/src/main/java/org/apache/iceberg/actions/compaction/DataCompactionStrategy.java

```diff
@@ -0,0 +1,103 @@
+
+public interface DataCompactionStrategy extends Serializable {
+
+  /**
+   * Returns the name of this compaction strategy
+   */
+  String name();
+
+  /**
+   * Returns a set of options which this compaction strategy can use. This is an allowed-list and any options not
+   * specified here will be rejected at runtime.
+   */
+  Set<String> validOptions();
+
+  default void validateOptions(Map<String, String> options) {
+    Sets.SetView<String> invalidOptions = Sets.difference(options.keySet(), validOptions());
+    if (!invalidOptions.isEmpty()) {
+      String invalidColumnString = invalidOptions.stream().collect(Collectors.joining(",", "[", "]"));
+      String validColumnString = validOptions().stream().collect(Collectors.joining(",", "[", "]"));
+
+      throw new IllegalArgumentException(String.format(
+          "Cannot use strategy %s with unknown options %s. This strategy accepts %s",
+          name(), invalidColumnString, validColumnString));
+    }
+  }
+
+  DataCompactionStrategy withOptions(Map<String, String> options);
+
+  /**
+   * Before the compaction strategy rules are applied, the underlying action has the ability to use this expression
+   * to filter the FileScanTasks which will be created when planning file reads that will later be run through
+   * this compaction strategy.
+   *
+   * @return an Iceberg expression to use when discovering file scan tasks
+   */
+  default Expression preFilter() {
+    return Expressions.alwaysTrue();
+  }
+
+  /**
+   * Removes all file references which this plan will not rewrite or change. Unlike the preFilter, this method can
+   * execute arbitrary code and isn't restricted to just Iceberg Expressions. This should be serializable so that
+   * Actions which run remotely can utilize the method.
+   *
+   * @param dataFiles iterator of live datafiles in a given partition
+   * @return iterator containing only files to be rewritten
+   */
+  Iterator<FileScanTask> filesToCompact(Iterator<FileScanTask> dataFiles);
+
+  /**
+   * Groups file scans into lists which will be processed in a single executable
+   * unit. Each group will end up being committed as an independent set of
+   * changes. This creates the jobs which will eventually be run as by the underlying Action.
+   *
+   * @param dataFiles iterator of files to be rewritten
+   * @return iterator of sets of files to be processed together
+   */
+  Iterator<List<FileScanTask>> groupsFilesIntoJobs(Iterator<FileScanTask> dataFiles);
+
+  /**
+   * Method which will actually rewrite and commit changes to a group of files
+   * based on the particular CompactionStrategy's Algorithm. This will most likely be
+   * Action framework specific.
+   *
+   * @param table table being modified
+   * @param filesToRewrite a group of files to be rewritten together
+   * @return a list of newly committed files
+   */
+  List<DataFile> rewriteFiles(Table table, List<FileScanTask> filesToRewrite);
```

Comment:
I think we need to consider the format v2's compaction strategy , at least in the API design level. Otherwise we will deprecate this API soon once we introduce compaction implementation for v2 in https://github.com/apache/iceberg/pull/2303/files

Reply:
For delete file I believe we should be relying on the underling framework's implementation of a the replace operator. For example the MergeInto command needs to have the same "Delete file" removal behavior as this particular rewrite pattern. That said, I still think you can fit just about any delete file compaction into this model as well since we are passing a list of FileScanTasks which should enumerate all positional delete files.  I think delete file comapctions, ie delete files merging and conversion between types should be a separate action.